### ISSUE TITLE
Same hyperparameter values with hp.choice

### ISSUE THREAD

**Issue by vh77** [hyperopt/hyperopt#370]
Hi,

When using hp.choice together with tpe.suggest, I notice the exact same set of values is tried out several times. Is that a normal behaviour? My search space is pretty large (I use ranges of integers to populate the set to be taken by hp.choice) but still, some values are almost never tried out and some are again and again.
Does TPE assume the function to minimize is not deterministic and that it makes sense to try several times the same set of hyperparameter values?

**Comment by vh77** [comment]
Same issue as here: https://github.com/hyperopt/hyperopt/issues/335

**Comment by thiagotei** [comment]
I'm having the same issue. Isn't the database supposed to catch data? Assuming that tpe thinks the tests are deterministic. 
Have you @vh77  tested whether specifying the search algorithm instead of tpe.suggest has a different outcome?

**Comment by vh77** [comment]
Hi Thiago,
I only tried tpe, I have to say. But then, if I go for random instead, I wouldn't be surprised to notice such a behaviour.
No really, I was not able to figure this out. Let me know if you have a clue... :)

**Comment by ashay88** [comment]
This issue #181 might be relevant

**Comment by Rubyr0id** [comment]
Experienced today this issue if with both hp.choice and hp.quniform.

hyperopt.__version__ ==  0.2

**Comment by xingyezhi** [comment]
@Rubyr0id maybe because the trials parameter, if number of  hp.choice less than trials, then it will repeat trials times. For example:
best = fmin(objscore,
        space=space,
        algo=hyperopt.random.suggest,
        max_evals=max_evals,
        trials=5)
hyopt({'nhidlayer': hp.choice('nhidlayer', [2])}
it will repeat 5 times even only one choice.

**Comment by Rubyr0id** [comment]
@xingyezhi my paramater space had uniform continuous distribution, so there were infinite number of points to try. I solved the issue by checking the trial.trials list of dictionaries inside my custom function which i had been minimizing - if current parameters are already contained in trials object the I immediately return {loss: None, status: STATUS_FAIL}

**Comment by ivan-marroquin** [comment]
Hi @xingyezhi ,

I liked the way how you avoid minimizing your test function by checking if current parameters were used previously. Could you share the way how did you do that?

Also, if I only return a loss: None without status. Will this influence the ability of the algorithm to find an optimal configuration of parameters?

Many thanks,
Ivan

**Comment by Rubyr0id** [comment]
@ivan-marroquin , assuming your question is addressed to me - my clumsy realization with two slightly different approaches:
`from hyperopt import hp, tpe, Trials, STATUS_OK, space_eval, STATUS_FAIL
from hyperopt.fmin import fmin
import numpy as np
from functools import partial

def loss_func(x1,x2):
    return x1**2+x2

# var1 - parse trials list on each iteration
# In case the None in 'loss' breaks your logic of processing 'trials' in later code, then just put there some very large value instead
def custom_objective_func(params):
    # for convenient comparison I transform dict to string
    joined_params = '.'.join(map(str,params.items()))
    # in case its first iteration
    if len(trials.trials)>1:
        for x in trials.trials[:-1]:
            space_point_index = dict([(key,value[0]) for key,value in x['misc']['vals'].items() if len(value)>0])
            if joined_params == '.'.join(map(str,space_eval(space,space_point_index).items())):
                return {'loss': None,'status': STATUS_FAIL}
    loss = loss_func(**params)
    return {'loss': loss ,'status': STATUS_OK}

# var2 - I assume it should be a bit faster, making use of an ability to store your own keys in output dictionary, which is appended to 'trials' by fmin
# Here, when dealing with already seen parameters, instead of None, I return 'loss' value, which was previously computed for this combination of parameters
def custom_objective_func(params):
    joined_params = '.'.join(map(str,params.items()))
    if len(trials.trials)>1:
        for x in trials.trials[:-1]:
            if joined_params == x['result']['joined_params']:
                loss = x['result']['loss']
                return {'loss': loss, 'status': STATUS_FAIL, 'joined_params':joined_params}
    loss = loss_func(**params)
    return {'loss': loss ,'status': STATUS_OK,'joined_params':joined_params}

max_evals = 50
init_evals = 5
trials=Trials()

space = {
        'x1': hp.choice('x1', np.arange(1,5,1, dtype=int)),
        'x2': hp.uniform('x2',0.01,70),
        }

best = fmin(fn=custom_objective_func,
            space=space,
            trials=trials,
            algo=partial(tpe.suggest,n_startup_jobs=init_evals),
            max_evals=max_evals,
            max_queue_len=1)

best_params = space_eval(space,best)
`

**Comment by Rubyr0id** [comment]
@ivan-marroquin , assuming your question is addressed to me - my code with two clumsy approaches:

from hyperopt import hp, tpe, Trials, STATUS_OK, space_eval, STATUS_FAIL
from hyperopt.fmin import fmin
import numpy as np
from functools import partial

def loss_func(x1,x2):
    return x1**2+x2

def custom_objective_func(params):
	'''
	1st approach:
	parse trials list on each iteration
	In case the None in 'loss' breaks your logic of processing 'trials' in later code, then just put there some very large value instead
	'''
    # for convenient comparison I transform dict to string
    joined_params = '.'.join(map(str,params.items()))
    # in case its first iteration
    if len(trials.trials)>1:
        for x in trials.trials[:-1]:
            space_point_index = dict([(key,value[0]) for key,value in x['misc']['vals'].items() if len(value)>0])
            if joined_params == '.'.join(map(str,space_eval(space,space_point_index).items())):
                return {'loss': None,'status': STATUS_FAIL}
    loss = loss_func(**params)
    return {'loss': loss ,'status': STATUS_OK}

def custom_objective_func(params):
	'''
	2nd approach:
	I assume it should be a bit faster, making use of an ability to store your own keys in output dictionary, which is appended to 'trials' by fmin 
	Here, when dealing with already seen parameters, instead of None, I return 'loss' value, which was previously computed for this combination of parameters
	'''
    joined_params = '.'.join(map(str,params.items()))
    if len(trials.trials)>1:
        for x in trials.trials[:-1]:
            if joined_params == x['result']['joined_params']:
                loss = x['result']['loss']
                return {'loss': loss, 'status': STATUS_FAIL, 'joined_params':joined_params}
    loss = loss_func(**params)
    return {'loss': loss ,'status': STATUS_OK,'joined_params':joined_params}

max_evals = 50
init_evals = 5
trials=Trials()

space = {
        'x1': hp.choice('x1', np.arange(1,5,1, dtype=int)),
        'x2': hp.uniform('x2',0.01,70),
        }

best = fmin(fn=custom_objective_func,
            space=space,
            trials=trials,
            algo=partial(tpe.suggest,n_startup_jobs=init_evals),
            max_evals=max_evals,
            max_queue_len=1)

best_params = space_eval(space,best)

**Comment by Rubyr0id** [comment]
@ivan-marroquin , assuming your question is addressed to me:

> I only return a loss: None without status. Will this influence the ability of the algorithm

Don't know for sure but I suppose the 'STATUS' key is there for a reason. Instead of just returning None, you can return 'loss' value, which was previously computed for this combination of parameters.
I believe this allows you to return only 'loss' key without 'status' with no risk of messing up the optimization algorithm

> Could you share the way how did you do that?

my code:

```
from hyperopt import hp, tpe, Trials, STATUS_OK, space_eval, STATUS_FAIL
from hyperopt.fmin import fmin
import numpy as np
from functools import partial

def loss_func(x1,x2):
    return x1**2+x2

def custom_objective_func(params):
    # in case its first iteration
    if len(trials.trials)>1:
        for x in trials.trials[:-1]:
            space_point_index = dict([(key,value[0]) for key,value in x['misc']['vals'].items() if len(value)>0])
            if params == space_eval(space,space_point_index):
                # instead of just returning None, you can return 'loss' value, which was previously computed for this combination of parameters
                # this allows you to return only 'loss' key without 'status' with no risk of messing up the optimization algorithm
                loss = x['result']['loss']
                return {'loss': None,'status': STATUS_FAIL}
    loss = loss_func(**params)
    return {'loss': loss ,'status': STATUS_OK}

max_evals = 50
init_evals = 5
trials=Trials()

space = {
        'x1': hp.choice('x1', np.arange(1,5,1, dtype=int)),
#         'x2': hp.choice('x2', np.arange(1,5,1, dtype=int)),
        'x2': hp.uniform('x2',0.01,70),
        }

best = fmin(fn=custom_objective_func,
            space=space,
            trials=trials,
            algo=partial(tpe.suggest,n_startup_jobs=init_evals),
            max_evals=max_evals,
            max_queue_len=1)

best_params = space_eval(space,best)
```

**Comment by xingyezhi** [comment]
@Rubyr0id，sorry, the parameter I given before is wrong, the parameter  should be max_evals.
What I mean is that, if you set a max_evals, it will run max_evals times anyway. So, maybe you can try to reduce the max_evals.
@ivan-marroquin , Rubyr0id have given his code. None without status, there is no influence, hyperopt will just ignore the parameter where status is not STATUS_FAIL.

**Comment by Rubyr0id** [comment]
@xingyezhi , I ran into this issue with repeating points when the algorithm was far from finding the optimal parameters, so reducing MAX_EVALS in this case could reduce the quality of "best found solution"

> it will run max_evals times anyway
True, but in my case the time hyperopt spend to decide the "next point to try" is notably smaller than the time my custom_loss_function computes (imagine you are fitting an XGBoost or NeuralNet on a large dataset for each combination of parameters), so each time I ran into the "already seen" parameters I just skip the most heavy part of the objective function
